### ISSUE TITLE
Self-documenting expression not included for f-string formatting

### ISSUE THREAD

**Issue by blhsing** [python/cpython#119489]
# Bug report

### Bug description:

Currently using a self-documenting expression with a fixed-width right-aligned specifier in an f-string does not actually align to the right as expected:
```python
foo = 1
foobar = 2
print(f'{foo=:>10}')
print(f'{foobar=:>10}')
```
which outputs:
```
foo=         1
foobar=         2
```
This behavior of a self-documenting expression not being included as part of the string formatting is not documented, and I think it should be considered a bug since it is more intuitive that the format specifier is applied to the entirety of the output.

One workaround to the issue is to format the self-documenting expression in a nested f-string first:
```python
foo = 1
foobar = 2
print(f'{f"{foo=}":>10}')
print(f'{f"{foobar=}":>10}')
```
which outputs:
```
     foo=1
  foobar=2
```
But perhaps a proper fix can format it as the following instead:
```
for=     1
foobar=  2
```

### CPython versions tested on:

3.12, 3.13

### Operating systems tested on:

Linux, Windows

**Comment by terryjreedy** [comment]
Not a bug, but a feature change request, and one which will certainly break existing code.
`f'{foo:10}'` formats the value of foo according to the spec.
`f'{foo=:10}' does the same thing after outputting the expression string plus '='.
I and I suspect most people would be surprised if adding '=' changes what got formatted.
If you want the expression text formatted, format it explicitly as desired.  Note that you suggested 2 equally sensible possibilities, and there are more.

[f-strings](https://docs.python.org/3/reference/lexical_analysis.html#formatted-string-literals) says
>To display both the expression text and its value after evaluation, (useful in debugging), an equal sign '=' may be added after the expression.

For debugging, the intended and hopefully short-term use, formatting the expression text is usually not needed.

You could try to get support for the idea on discuss.py ideas, but breaking changes require substantial reason beyond that given here.  So I suggest not trying.

**Comment by ericvsmith** [comment]
I don't see how this could work. The value being formatted isn't necessarily a string, so the format spec might not make sense for the final string value being produced. Consider:
```pycon
>> import datetime
>>> f'The year is {datetime.datetime.now()=:%Y}'
'The year is datetime.datetime.now()=2024'

```
How would %Y be interpreted for the entire string?

**Comment by blhsing** [comment]
> I don't see how this could work. The value being formatted isn't necessarily a string, so the format spec might not make sense for the final string value being produced. Consider:
> 
> ```
> >> import datetime
> >>> f'The year is {datetime.datetime.now()=:%Y}'
> 'The year is datetime.datetime.now()=2024'
> ```
> 
> How would %Y be interpreted for the entire string?

Self-documenting expression is meant for debugging output, where the returning values of `__str__` or `__repr__` are generally preferred over custom formats.

In your example, the output of `datetime.datetime.now()=2024` is rather confusing because `datetime.datetime.now()` certainly isn't `2024`, and if you really need just the year as the debugging output you would use f'{format(datetime.datetime.now(), "%Y")=}' for an output of `format(datetime.datetime.now(), "%Y")='2024'` so that the meaning of the output is much clearer.

In other words, the idea can work by mandating that a self-documenting expression is always converted to a string (unless it's int or float) before the format specifier is applied. It just makes the usage more intuitive.

@terryjreedy Thanks for the feedbacks. I will repost this as an idea in Discourse. The idea is for easier reading of debugging output after all, and as debugging output I don't think breaking changes can be that critical, so the idea should still be worth a consideration.

**Comment by ericvsmith** [comment]
> In other words, the idea can work by mandating that a self-documenting expression is always converted to a string (unless it's int or float) before the format specifier is applied. It just makes the usage more intuitive.

That would be a huge compatibility break. This feature is used far beyond just interactive debugging. I personally have code in production that this proposed change would break.

I'm going to close this, pending any discussion on Discourse.
